### PR TITLE
feat: added filter by cell and gene count. #1820, #1823

### PR DESCRIPTION
**Pull request by MillenniumFalconMechanic**
### Reviewers
**Functional:** 
@tihuan 

---

## Changes
- Added filter by cell count ([mocks](https://www.figma.com/file/18a4QULVIfpFQhTn1FGh8a/Filtering?node-id=1468%3A9713)).
- Added filter by gene count ([mocks](https://www.figma.com/file/18a4QULVIfpFQhTn1FGh8a/Filtering?node-id=1468%3A9713)).

#### Cell Count
* Cell count filter displayed in datasets filter sidebar.
* Slider UI matches [mocks](https://www.figma.com/file/18a4QULVIfpFQhTn1FGh8a/Filtering?node-id=1468%3A9713).
* Selected state matches [mocks](https://www.figma.com/file/18a4QULVIfpFQhTn1FGh8a/Filtering?node-id=2681%3A31105).
* Selecting min and max cell count values filters corresponding datasets result set.

#### Gene Count
* Gene count filter displayed in datasets filter sidebar.
* Slider UI matches [mocks](https://www.figma.com/file/18a4QULVIfpFQhTn1FGh8a/Filtering?node-id=1468%3A9713).
* Selecting min and max gene count values filters corresponding datasets result set.

### CODE REVIEW

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx [hunk 1]
```diff
@@ -328,6 +328,8 @@ export default function GeneSearchBar({ onGenesChange }: Props): JSX.Element {
           }
           renderOption={(option) => option.name}
           onPaste={handlePaste}
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment -- TODO revisit lint errors
+          // @ts-ignore -- TODO revisit lint errors
```
Comment (MillenniumFalconMechanic) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx:332]: I was getting an error here when running `npm run lint`.

File: frontend/src/common/hooks/useCategoryFilter.ts [hunk 2]
```diff
@@ -462,15 +593,93 @@ function isFilterEqual<T extends Categories>(
...
+): categoryValue is KeyedSelectCategoryValue {
+  return (categoryValue as KeyedSelectCategoryValue).get !== undefined;
+}
+
+/**
+ * Update selected state of categories to match the current set of selected filters.
+ * @param nextFilterState - Filter state being built on select of filter.
+ * @param filters - Current set of selected category values (values) or ranges keyed by category (id).
+ */
+function setSelectedStates<T extends Categories>(
```
Comment (MillenniumFalconMechanic) [inline frontend/src/common/hooks/useCategoryFilter.ts:634]: Refactored (and renamed) [flagSelectedCategoryValues()](https://github.com/chanzuckerberg/single-cell-data-portal/pull/2017/files#diff-e6d68019c2a59230f504bd6a5d662355f67c0d10c25fadc0823bfc27d7815d5aL387) to handle range functionality.

File: frontend/src/common/hooks/useCategoryFilter.ts [hunk 3]
```diff
@@ -504,54 +718,53 @@ function summarizeCategories<T extends Categories>(
...
-          accum.set(categoryValueKey, categoryValue);
-        }
-        // Increment category value count.
-        categoryValue.count++;
-      });
-      return accum;
-    },
-    new Map<CategoryValueKey, CategoryValue>()
-  );
+    // Init category value it doesn't already exist. Default selected state to false (selected state is updated
```
Comment (tihuan) [inline frontend/src/common/hooks/useCategoryFilter.ts:751]: Potential addition 😆 ?
```suggestion
    // Init category value if it doesn't already exist. Default selected state to false (selected state is updated
```

File: frontend/src/components/common/Filter/common/utils.ts [hunk 4]
```diff
@@ -10,6 +10,31 @@ import {
...
 
+/**
+ * Number of digits to format decmial points to.
+ */
+const FIXED_TO = 1;
+
+/**
+ * Value to divide numbers by when scaling by million.
+ */
+const SCALE_MILLION = 1000000;
```
Comment (tihuan) [inline frontend/src/components/common/Filter/common/utils.ts:21]: for readability 😄 !

```suggestion
const SCALE_MILLION = 1_000_000;
```

File: frontend/src/components/common/Filter/common/utils.ts [hunk 5]
```diff
@@ -22,3 +47,42 @@ export function ontologyCellAccessorFn(
...
 }
+
+/**
+ * Format number to the correct scale and possibly round.
+ * @param num - Number to format.
+ * @returns String representation of given number, possibly rounded, with corresponding scale symbol.
+ */
+export function formatNumberToScale(num: number): string {
+  // Return numbers less than 1000 rounded.
+  if (num < 1000) {
```
Comment (tihuan) [inline frontend/src/components/common/Filter/common/utils.ts:58]: should we use the constants here?

E.g., `num < SCALE_THOUSAND`, `num < 10 * SCALE_THOUSAND`, `num < SCALE_MILLION`, etc.

File: frontend/src/components/common/Filter/common/utils.ts [hunk 6]
```diff
@@ -22,3 +47,42 @@ export function ontologyCellAccessorFn(
...
+
+/**
+ * Round up the given number to the given fixed digits, removing any insignificant decimals.
+ * @param num - Number to round to significant digits.
+ * @param scale - Number to divide given number by.
+ * @param fixTo - Number of digits to fix number to.
+ * @returns Rounded number to the given fixed digits.
+ */
+function scaleAndFix(num: number, scale: number, fixTo: number): number {
+  return parseFloat((num / scale + Number.EPSILON).toFixed(fixTo));
```
Comment (tihuan) [inline frontend/src/components/common/Filter/common/utils.ts:87]: never seen this before! Really cool 😮

File: frontend/src/components/common/Filter/index.tsx [hunk 7]
```diff
@@ -19,29 +24,37 @@ interface Props {
...
-                onFilter={onFilter}
-                onUpdateSearchValue={onUpdateSearchValue}
-                values={values}
-              />
+              isSelectCategoryView(categoryView) ? (
+                <FilterMenu
+                  categoryKey={key}
+                  filterCategoryValues={filterCategoryValues}
+                  filterCategoryValuesWithCount={filterCategoryValuesWithCount}
+                  isMultiselect={true} // Can possibly be single select with future filter types
```
Comment (tihuan) [inline frontend/src/components/common/Filter/index.tsx:39]: Abbreviation!
```suggestion
                  isMultiselect // Can possibly be single select with future filter types
```

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx (hunk 1 above)
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx:332]: Ah sorry for the trouble! I have a [fix](https://github.com/chanzuckerberg/single-cell-data-portal/pull/2018) for this after an SDS PR is merged later today!

**Review by tihuan — APPROVED** [review]
LGTM! Looks super good 🎉 Thank you 👏 

Only some super tiny nits I could find 😆

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx (hunk 1 above)
Comment (MillenniumFalconMechanic) [reply frontend/src/views/WheresMyGene/components/GeneSearchBar/index.tsx:332]: No worries at all! Thanks for the update, Timmy.

**Pull request merged**